### PR TITLE
Update Implementation for Date Creation in DG

### PR DESCRIPTION
**Pull request by pyk595**
The Developer Guide does not have documentation for date creation
processes. This may make the design confusing for reviewers.

View [here](https://github.com/pyk595/tp/blob/branch-update-DG/docs/DeveloperGuide.md)

Let's do the following to close AY2122S2-CS2103T-T17-3#174:
- Update implementation for date creation
- Add design consideration
- Update glossary

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -137,6 +137,19 @@ The `Storage` component,
 * inherits from both `AddressBookStorage` and `UserPrefStorage`, which means it can be treated as either one (if only the functionality of only one is needed).
 * depends on some classes in the `Model` component (because the `Storage` component's job is to save/retrieve objects that belong to the `Model`)
 
+In the event that data saved in the saved json file is corrupted, the `Storage component` will make a copy of the
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:140]: 1. "corrupted or can't be read".
2. `Storage` component

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -137,6 +137,19 @@ The `Storage` component,
...
+existing json file.
+
+<img src="images/ReadAddressBookSequenceDiagram.png" width="550" />
+
+After the `MainApp` calls `JsonAddressBookStorage#readAddressBook()`, if the data is corrupted, `JsonAddressBookStorage`
+will call the `backup()` method and make a copy of the `addressbook.json`. The copied file will be saved in the `data`
+directory in a file called `backup.json`.
+
+Users will still be able to use all features available, but the existing `addressbook.json` will be overwritten.
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:149]: "overwritten with the default contact list"

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -256,7 +269,103 @@ as values.<br>
   * Cons: Updating `UniqueTagList` takes O(logn) time every time; requires additional data structure to maintain
   `UniqueTagList` accurately.
 
---------------------------------------------------------------------------------------------------------------------
+### Date Features
+
+#### Implementation
+
+As seen in the [`Model` component](#model-component), Each `Person` object contains its own `BirthDate` object,
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:276]: "each"

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+a `ContactedInfoList` and a `ReminderList`. The `ContactedInfoList` stores `ContactedInfo` objects,
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:277]: list of `ContactedInfo`? since there's no class, `ContactedInfoList`

File: docs/DeveloperGuide.md [hunk 5]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+objects need to be displayed in the same format for consistency, we have a `DocumentedDate` parent class to ensure that 
+the three “date” type objects will use the same `toString` method in `DocumentedDate` to allow users to display the 
+dates in the same format, as seen in the class diagram below.
+
+<img src="images/DocumentedDateClassDiagram.png" width="450" />
+
+##### Documented Date
+
+`DocumentedDate` objects have the following characteristics:
+* A `DocumentedDate` object has a private `LocalDate` member to allow the application to easily display a formatted date.
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:289]: "to easily display a formatted date" -> "to store a date"?

File: docs/DeveloperGuide.md [hunk 6]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+
+
+##### Documented Date Child Classes
+
+Despite inheriting from the `DocumentedDate` parent class, these 3 “date” type objects have different behaviours. 
+A `BirthDate` object needs to be recurring, to check if the person’s birthday is occurring on the same day despite 
+being saved in a past year. A `RecentDate` needs to be a date that occurs in the past, and a `ReminderDate` needs to be 
+a date that has not yet occurred. To model this more concretely, we implement some checks using the `ParserUtil` class.
+
+
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:301]: extra line break?

File: docs/DeveloperGuide.md [hunk 7]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+The three "date" type objects, are primarily created using static methods in the `ParserUtil` class. 
+However, there are public constructors to create each `BirthDate`, `RecentDate` and `ReminderDate` object. This is to 
+enable better testability.
+
+<img src="images/BirthDateCreationSequenceDiagram.png" width="450" />
+
+ As seen in the sequence diagram above, which shows the process of creating a `BirthDate` object.
+ `ParserUtil#parseBirthDate(validDate)` is called, with the user supplying a `validDate` in the form of a String.
+The method will process the `validDate` into a `String` called `trimmedDate` and call
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:310]: "process" -> "trim"

File: docs/DeveloperGuide.md [hunk 8]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+The three "date" type objects, are primarily created using static methods in the `ParserUtil` class. 
+However, there are public constructors to create each `BirthDate`, `RecentDate` and `ReminderDate` object. This is to 
+enable better testability.
+
+<img src="images/BirthDateCreationSequenceDiagram.png" width="450" />
+
+ As seen in the sequence diagram above, which shows the process of creating a `BirthDate` object.
+ `ParserUtil#parseBirthDate(validDate)` is called, with the user supplying a `validDate` in the form of a String.
+The method will process the `validDate` into a `String` called `trimmedDate` and call 
+ `DocumentedDate#isValidDate(trimmedDate)` to check if it is a valid date. if the `trimmedDate` is indeed a valid date,
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:311]: "it is in a valid date format"

File: docs/DeveloperGuide.md [hunk 9]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+when used, instead of an invalid or unexpected entry.
+
+#### Design Consideration
+
+##### Aspect: How date type objects are created
+
+* Alternative 1 (current implementation): "Date" type objects extend `DocumentedDate`.<br>
+    * Pros: We can use polymorphism to modify certain behaviour of each specific child class (i.e. Allowing `BirthDate` 
+      objects to be read as recurring dates when necessary.)
+    * Less bug prone as each child class is independent of each other.
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:344]: is the bullet intentional? check for other bullets in `Design Consideration` as well.

File: docs/DeveloperGuide.md [hunk 10]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+Having two levels of checks ensure that when the working "date" type objects are less bug prone due when using the specific
+"date" type object for their designated usages. This makes features such as `after` or `within` return valid entries
+when used, instead of an invalid or unexpected entry.
+
+#### Design Consideration
+
+##### Aspect: How date type objects are created
+
+* Alternative 1 (current implementation): "Date" type objects extend `DocumentedDate`.<br>
+    * Pros: We can use polymorphism to modify certain behaviour of each specific child class (i.e. Allowing `BirthDate`
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:342]: I think this is not polymorphism, which is like you are using `DocumentedDate` as the object instead. and you say the object takes many forms, like ReminderDate...

File: docs/DeveloperGuide.md [hunk 11]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+#### Design Consideration
+
+##### Aspect: How date type objects are created
+
+* Alternative 1 (current implementation): "Date" type objects extend `DocumentedDate`.<br>
+    * Pros: We can use polymorphism to modify certain behaviour of each specific child class (i.e. Allowing `BirthDate` 
+      objects to be read as recurring dates when necessary.)
+    * Less bug prone as each child class is independent of each other.
+    * Cons: More overhead as more classes are required, possibly more coupling if code is not written with 
+      Liskov Substitution Principle in mind.
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:346]: LSP: you might get bashed. `isToday()` in `BirthDate` violates LSP.

File: docs/DeveloperGuide.md [hunk 12]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:347]: unnecessary line break?

File: docs/DeveloperGuide.md [hunk 13]
```diff
@@ -256,7 +269,103 @@ as values.<br>
...
+* Alternative 1 (current implementation): "Date" type objects extend `DocumentedDate`.<br>
+    * Pros: We can use polymorphism to modify certain behaviour of each specific child class (i.e. Allowing `BirthDate` 
+      objects to be read as recurring dates when necessary.)
+    * Less bug prone as each child class is independent of each other.
+    * Cons: More overhead as more classes are required, possibly more coupling if code is not written with 
+      Liskov Substitution Principle in mind.
+   
+ 
+* Alternative 2: Use `DocumentedDate`, but use other components to differentiate (i.e. `enum` types)
```
Comment (ngjunkang) [inline docs/DeveloperGuide.md:349]: elaborate on this? I don't really understand. Use `DocumentedDate` as the actual date?

**Review by ngjunkang — CHANGES_REQUESTED** [review]
Some changes to be made

File: docs/DeveloperGuide.md (hunk 6 above)
Comment (ngjunkang) [reply docs/DeveloperGuide.md:301]: nice

File: docs/DeveloperGuide.md [hunk 14]
```diff
@@ -256,7 +269,104 @@ as values.<br>
...
+
+<img src="images/ReminderDateCreationSequenceDiagram.png" width="450" />
+
+As seen above, the process exactly mirrors that of the `BirthDate` object. However, the biggest difference is that the
+check using `getDaysPassed()` must make sure that `ReminderDate` objects cannot be created with a date that has passed.
+This is because the function of the reminder is to ensure that users can keep track of tasks that have not yet been done
+or important events that are upcoming.
+
+The `RecentDate` objects are much more similar to the `BirthDate` objects as they have the same requirement of not being
+able to be created using a date in the future. The rationale is because `RecentDate` keep track of the user's
```
Comment (cashewnade) [inline docs/DeveloperGuide.md:330]: `RecentDate` keeps track of...

File: docs/DeveloperGuide.md [hunk 15]
```diff
@@ -256,7 +269,104 @@ as values.<br>
...
+interaction with clients and contacts, and the user must have interacted with the contact before they save the
+interaction record in the application.
+
+Having two levels of checks ensure that when the working "date" type objects are less bug prone due when using the specific
```
[the remark this answers is not in the document]
Comment (cashewnade) [reply docs/DeveloperGuide.md:334]: "Having two levels of checks ensures that working "date" type objects are less bug prone, when using specific "date" type objects for their designated usages."

**Review by ngjunkang — APPROVED** [review]
Looks good..

**Pull request merged**